tech-specs/src/specs/ledger.md

```diff
@@ -0,0 +1,190 @@
+# The ledger
+
+## The protocol
+
+- TODO describe DKG transactions
+
+### Transactions
+
+A transaction [encoded with proto3](./encoding.md#transactions) received from ABCI `DeliverTx` method is executed in two main steps:
+
+1. [Transaction execution](#transaction-execution)
+1. [Validity predicates check](#validity-predicates-check)
+
+#### Transaction execution
+
+For any error encountered in any of the following steps of transaction execution, the protocol MUST charge the gas used by the transaction and discard any storage changes that the transaction attempted to perform.
+
+1. Charge a base transaction [gas](#gas): \\[ \verb|BASE_TRANSACTION_FEE| \\]
+1. Decode the transaction bytes and validate the data.
+1. Charge WASM compilation gas, proportional to the bytes `length` of the `code` field of the transaction (this is because the WASM code is compiled with a single-pass compiler): \\[ \verb|length| * \verb|COMPILE_GAS_PER_BYTE| \\].
+1. [Validate the WASM code](#wasm-validation) from the `code` field of the transaction.
+1. Inject a [gas counter](#gas) into the `code`.
+1. Inject a [stack height](#stack-height-limiter) limiter into the `code`.
+1. Compile the transaction `code` with single-pass compiler. The compilation computational complexity MUST be linear in proportion to the `code` size.
```

Comment:
Can we link to something describing the single-pass compiler?

Reply:
The best I could find was this blog post describing it https://medium.com/wasmer/a-webassembly-compiler-tale-9ef37aa3b537
I've added the link